src/Networking/NexusMods.Networking.NexusWebApi/RunUpdateCheck.cs

```diff
@@ -49,28 +49,56 @@ public static async Task<PerFeedCacheUpdaterResult<PageMetadataMixin>> CheckForM
     public static async Task UpdateModFilesForOutdatedPages(IDb db, ITransaction tx, ILogger logger, INexusGraphQLClient gqlClient, PerFeedCacheUpdaterResult<PageMetadataMixin> result, CancellationToken cancellationToken)
     {
         // Note(sewer): Undetermined items may be removed items from the site; or
-        // caused by programmer error, so wr should log these whenever possible,
+        // caused by programmer error, so we should log these whenever possible,
         // but they should not cause a critical error; in case it's simply the result
         // of mod removal such as DMCA takedown.
-        foreach (var mixin in result.UndeterminedItems)
+        using var semaphore = new SemaphoreSlim(16);
+        var tasks = new List<Task>();
+
+        // Helper function to process a single mixin with error handling
+        async Task ProcessMixin(PageMetadataMixin mixin, bool isUndetermined)
         {
             try
             {
-                await UpdateModPage(db, tx, gqlClient, cancellationToken, mixin);
+                await semaphore.WaitAsync(cancellationToken);
+                try
+                {
+                    await UpdateModPage(db, tx, gqlClient, cancellationToken, mixin);
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
```

Comment:
`semaphore.WaitAsync(cancellationToken);` can actually return without acquiring the resource if the token is already cancelled on first enter, which can lead to a release of something that was never acquired in the `finally` block.

`WaitAsync(cancellationToken);` returns false if the resource is never acquired, so you can use that do determine whether you need to release or not in the finally.

Reply:
Yeah, right I forgot about the specific overload thing